rfcs/0120-artifact-service.md

```diff
@@ -220,11 +230,80 @@ objsvc <-- s3 200 ETag {etag: 123456}
 uploader <-- objsvc 200
 ```
 
-### Deletetion
+### Part Size
+```
+GET /upload-info
+GET /upload-info/:origin
+```
+
+Multipart uploads in various backing services have differing requirements related
+to the size of each individual part.  This endpoint will return information about
+part sizes for a given origin.  In the case that the origin is not specified, the
+origin of the requester will be used.
+
+The response of this method for an S3 origin will be:
+
+```
+[
+  {
+    type: 'single-part',
+    max_size: 5 * 1024 * 1024,
+    min_size: 0,
+    max_part_count: 1,
+    min_part_count: 1,
+    min_part_size: 5 * 1024 * 1024,
+    max_part_size: 5 * 1024 * 1024,
+    part_size_multiple: 1,
+  },
+  {
+    type: 'multipart',
+    max_size: 5 * 1024 * 1024 * 1024,
+    min_size: 0,
+    max_part_count: 10000,
+    min_part_count: 1,
+    min_part_size: 5 * 1024 * 1024,
+    max_part_size: 5 * 1024 * 1024,
+    part_size_multiple: 1,
+  },
+]
+```
+
+All services will use a shape like this, changing their values to reflect their
+reality.  This format will give the uploader the ability to choose from the
+supported types of uploads.  Only those upload types which work with this
+upload flow will be supported, regardless of whether more options are available
+in the underlying API.
+
+The keys for this response will have the following meaning:
+
+* `type`: an identifier for the style of upload, unique to each underlying
+  service
+* `min`/`max_size`: The size range for the overall object in bytes
+* `min`/`max_part_count`: The range of number of parts allowed
+* `min`/`max_part_size`: The size range for each part in bytes
+* `part_size_multiple`: If a service requires parts be a multiple of a number,
+  this is the value.
+
+The ranges must all work together such that the most strict condition is not
+exceeded.  For example if the maximum number of parts is 10,000 and you want to
+use 5MB parts, you are limited to 5MB * 10,000, not 5TB.
+
+Of note is that the minimum part size is interpreted slightly differently.  In
+order to support last part part sizes which are not the minimum part size, the
+last part may be smaller than the minimum value specified here.
+
+This endpoint provides advisory information and does not obviate the service from
+checking these values to be valid when the object is created.  The object creation
+endpoint must always validate its provided values.
+
+The value of this URL is cachable for at least 30 minutes, which implies that
+origin to backend mapping must be stable for 30 minutes at a time.
```

Comment:
This means the service will need a way to stop advertising an origin/backend pair, but still allow requests for it, right?  Basically we'd set `advertised: false`, wait 30 minutes, then delete the entry from the config?

Reply:
We think we don't need to worry about this any more after meeting about part sizes